File: lab4/lab4.py
# Russell Todd
# Lab 4 for UWYO COSC 2030
# Stacks and Queues
#
# You'll be checking parentheses and reversing strings just like you did in C++
# However, in Python the class you'll be using is a List

paren1 = "(()((((()()()((()(((()()()()(((()(()()()(())()())()()))))()()()))()))()())())())))";
paren2 = "()((((((()(((((()((()()()()()(()))))))))()))()(())(())(((((()()(())))(()()())(()))";
paren3 = "(()((()(()()(()(((((()()(()()()((((()()(()()))()))))()()())))))()()())()()))())())";
paren4 = "(()()(((()()(()(()()(()()()()()()(()(((((((((())())))))()))))()()))()())()()))()))";
paren5 = "((())";
paren6 = ")))(((";

rev1 = "sekopog";
rev2 = "racecar";
rev3 = "regnolsignirtssiht";
rev4 = "wonybnwodsihtevahyllufepohdluohssyuguoy";

parens = [paren1, paren2, paren3, paren4, paren5, paren6]
rever = [rev1, rev2, rev3, rev4]

# Finish this function
# You must complete it using a List in a similar manner as a Queue
def parenCheck(toCheck):
    pseudoQueue = []
    counter = 0

    for char in toCheck:
        pseudoQueue.insert(0, char)

    while (pseudoQueue):
        if (pseudoQueue[-1] == "("):
            counter += 1
        elif (pseudoQueue[-1] == ")"):
            counter -= 1
            if counter < 0:
                return False
            
        pseudoQueue.pop()

    return True if counter == 0 else False


# Finish this function
# You must complete it using a List in a similar manner as a Stack
def stringReverse(toReverse):
    pseudoStack = []
    returnString = ""

    for char in toReverse:
        pseudoStack.append(char)
    
    while (pseudoStack):
        returnString += pseudoStack.pop()

    return returnString


# Time to check some parentheses
for i in range(6):
    print("Checking string"+str(i+1))
    if parenCheck(parens[i]):
        print("String"+str(i+1)+" is good.")
    else:
        print("String"+str(i+1)+" is bad.")

# Time to reverse some strings
for j in range(4):
    print("Reversing string"+str(j+1))
    print(stringReverse(rever[j]))

File: lab4/lab4.cpp
//UWYO COSC 2030
//Lab 4
//Written by Danny Radosevich
//8-4-19

#include <iostream>
#include <stack>
#include <queue>
#include <vector>

using namespace std;


//function declarations
string stringReverse(string toReverse);
bool parenCheck(string toCheck);

/*DO NOT CHANGE MAIN OR THE FUNCTION DECLARATIONS*/
int main()
{
  //paren strings
  string paren1 = "(()((((()()()((()(((()()()()(((()(()()()(())()())()()))))()()()))()))()())())())))";
  string paren2 = "()((((((()(((((()((()()()()()(()))))))))()))()(())(())(((((()()(())))(()()())(()))";
  string paren3 = "(()((()(()()(()(((((()()(()()()((((()()(()()))()))))()()())))))()()())()()))())())";
  string paren4 = "(()()(((()()(()(()()(()()()()()()(()(((((((((())())))))()))))()()))()())()()))()))";
  string paren5 = "((())";
  string paren6 = ")))(((";

  string rev1 = "sekopog";
  string rev2 = "racecar";
  string rev3 = "regnolsignirtssiht";
  string rev4 = "wonybnwodsihtevahyllufepohdluohssyuguoy";

  //Ignore this for now, these vectors are just part of the program
  vector<string> parens;
  parens.push_back(paren1);
  parens.push_back(paren2);
  parens.push_back(paren3);
  parens.push_back(paren4);
  parens.push_back(paren5);
  parens.push_back(paren6);

  vector<string> rever;
  rever.push_back(rev1);
  rever.push_back(rev2);
  rever.push_back(rev3);
  rever.push_back(rev4);

  //call the parenCheck
  for(int i=0; i<6; i++)
  {
    cout<<"Checking string"<<i+1<<endl;
    if(parenCheck(parens.at(i)))
    {
      cout<<"String"<<i+1<<" is good"<<endl<<endl;
    }
    else
    {
      cout<<"String"<<i+1<<" is bad"<<endl<<endl;
    }
  }

  //now to reverse some strings
  for(int i =0; i<4; i++)
  {
    cout<<"Reversing string"<<i+1<<endl;
    cout<<stringReverse(rever.at(i))<<endl<<endl;
  }
}

//This function takes a string and reverses it using a stack
//it returns the reversed string
string stringReverse(string toReverse)
{
  stack<char> reverse; //declare the stack
  string newString = "";
  
  for (int chr=0; chr < toReverse.length(); chr++) {
    reverse.push(toReverse[chr]);
  }

  while (!reverse.empty()) {
    newString += static_cast<char>((reverse.top()));
    reverse.pop();
  }


  return newString;
}

//this function uses a queue to check if a string of parens is properly matched
bool parenCheck(string toCheck)
{
  queue<char> checker; //declare my queue
  int counter=0;

  for (int chr=0; chr < toCheck.length(); chr++) {
    checker.push(toCheck[chr]);
  }

  while (!checker.empty()) {
    if (checker.front() == '(') {
        counter += 1;
    } else if (checker.front() == ')') {
        counter -= 1;

        if (counter < 0) {
            return false;
        }
    }

    checker.pop();
  }
  return !(counter); //need a return statement to compile
}